Augment Python PlanningSceneInterface API

rhaschke opened the pull request:
This PR addresses several shortcomings of the current python API for PSI. See individual commit messages for details.
Closes #3527

v4hn commented on moveit_commander/src/moveit_commander/planning_scene_interface.py:
```diff
@@ -127,36 +127,54 @@ def add_plane(self, name, pose, normal=(0, 0, 1), offset=0):
+            object = AttachedCollisionObject(object=object)
+
+        object.link_name = link
+        if touch_links is not None:
+            object.touch_links = touch_links if touch_links is not None else [link]
```
```suggestion
        object.touch_links = touch_links if touch_links is not None else [link]
```

v4hn commented on moveit_commander/src/moveit_commander/planning_scene_interface.py:
```diff
@@ -127,36 +127,54 @@ def add_plane(self, name, pose, normal=(0, 0, 1), offset=0):
-    def attach_object(self, attached_collision_object):
-        """Attach an object in the planning scene"""
-        self.__submit(attached_collision_object, attach=True)
+    def attach_object(self, object, link, touch_links=None):
+        """Attach an object to the given link"""
+        if isinstance(object, str):
+            object = self.__make_existing(object)
+        if isinstance(object, CollisionObject):
+            object = AttachedCollisionObject(object=object)
+
+        object.link_name = link
```
```suggestion
     def attach_object(self, object, link=None, touch_links=None):
        """Attach an object to the given link"""
        if isinstance(object, str):
            object = self.__make_existing(object)
        if isinstance(object, CollisionObject):
            object = AttachedCollisionObject(object=object)

    if link is not None:
        object.link_name = link
```

Looks like github does not support scattered diffs in the suggestions -.-
Anyway, I propose to support passing an `AttachedCollisionObject` which already specifies the link in the message.

v4hn commented on moveit_commander/src/moveit_commander/planning_scene_interface.py:
```diff
@@ -127,36 +127,54 @@ def add_plane(self, name, pose, normal=(0, 0, 1), offset=0):
+        if touch_links is not None:
+            object.touch_links = touch_links if touch_links is not None else [link]
+
+        self.__submit(object, attach=True)
```
Why not get rid of the attach parameter here altogether? The current code will break with an `AttachedCollisionObject` and `attach == False` and vice versa, right? So we might as well check via `isinstance` as you do above.

rhaschke replied:
:+1:

rhaschke replied:
The attach argument is needed for __submit to actually add an attached object. There is no type-checking in that function.

v4hn replied:
Yes, there isn't. I rather suggested to add it there. I see why it wasn't there originally in the "never check types" spirit, but then providing the type as a boolean argument isn't much better...

rhaschke replied:
Please feel free to augment my PR according to your idea. I still don't get it in full detail and I cannot assess the consequences of this change right now. At least, it is an API change - but of a private function. So, that shouldn't pose a big problem.
Sorry, I'm still somewhat dizzy due to my illness...

v4hn commented on moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h:
```diff
@@ -123,6 +123,9 @@ class PlanningSceneInterface
   bool
   applyAttachedCollisionObjects(const std::vector<moveit_msgs::AttachedCollisionObject>& attached_collision_objects);
 
+  /** \brief Get given components from move_group's PlanningScene */
```
```suggestion
  /** \brief Get given components from move_group's PlanningScene *
    * \param components bitfield as specified in moveit_msgs::PlanningSceneComponents */
```

v4hn commented on moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp:
```diff
@@ -228,6 +228,19 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
+    request.components.components = components;
+    if (!planning_scene_service_.call(request, response))
+    {
+      ROS_WARN_NAMED(LOGNAME, "Could not call planning scene service");
```
```suggestion
      ROS_ERROR_STREAM_NAMED(LOGNAME, "Could not call planning scene service '" << planning_scene_service_.getService() << "'");
```

v4hn commented on moveit_commander/src/moveit_commander/planning_scene_interface.py:
```diff
@@ -380,3 +381,38 @@ def make_plane(self, name, pose, normal=(0, 0, 1), offset=0):
+
+
+AllowedCollisionMatrix.set_default = acm_set_default
+AllowedCollisionMatrix.set_allowed = acm_set_allowed
```
Can you please comment on how usual/reasonable you think these monkey patches are?
I've never seen anyone patch ROS message objects before (granted, I didn't look for it) and at first glance it looks useful, but I also think it's a can of worms I'd prefer closed...

rhaschke replied:
This simply augments the standard ROS-message class `AllowedCollisionMatrix` with two new functions.

v4hn replied:
I know what it does. I'm not a fan of monkey patching, but of course the resulting interface looks nice when used.
I don't know how roundabout it would be to add it directly in moveit_msgs instead of here for example.

rhaschke replied:
I am not aware of a possibility to add methods directly in moveit_msgs. Of course, this would be even nicer. If you know an option to do so, please implement that in a follow-up PR.

v4hn reviewed (changes requested):
Great simplifications and better functionality, yay.

I wrote all suggestions on the website, so I guess format issues and typos are to be expected.

rhaschke commented:
I hope I addressed all your comments, @v4hn. Sorry for the cluttered replies to your comments. I used VS code for this and it didn't assign them to the original locations.

rhaschke commented:
@v4hn, is this ready to get merged? The additional changes you asked for, go far beyond this PR.

v4hn commented:
You are right. I still don't like the monkey patches, but they are definitely useful and I don't see a better way to provide the methods.

I'm not sure whether you want to clean up the commits before merging or merge as squash, so I'll leave it to you.
